### ISSUE TITLE
`rye sync` with uv enabled misses numpy dependency when installing pandas

### ISSUE THREAD

**Issue by Zaloog** [astral-sh/rye#875]
### Steps to Reproduce

1. `rye init project && cd project`
2. `rye pin 3.11.8`
3. `rye add pandas`
4. `rye sync` (not needed if autosync enabled)

Config:
```toml
[default]
# This is the default value that is written into new pyproject.toml
# files for the `project.requires-python` key
requires-python = ">= 3.8"

# This is the default toolchain that is used
toolchain = "cpython@3.11.8"

[behavior]
# Enables global shims when set to `true`.  This means that the installed
# `python` shim will resolve to a Rye managed toolchain even outside of
# virtual environments.
global-python = true

# When set to `true` enables experimental support of uv as a replacement
# for pip-tools. Learn more about uv here: https://github.com/astral-sh/uv
use-uv = true
```

### Expected Result

Pandas and all the dependencies get installed into the .venv

### Actual Result

pandas is added to `pyproject toml` and rye installs the packages. the numpy dependency is missing though
![image](https://github.com/astral-sh/rye/assets/57259263/517dc34e-bc38-40d0-b5a4-74062eeb013b)

Compared to result when pandas is installed with `uv`

![image](https://github.com/astral-sh/rye/assets/57259263/b2a3b6ef-8020-4dc0-92b7-e2a9535376e2)

### Version Info

```
rye 0.29.0
commit: 0.29.0 (42b179f36 2024-03-11)
platform: windows (x86_64)
self-python: cpython@3.12.2
symlink support: false
uv enabled: true
```

**Comment by bluss** [comment]
Autosync is enabled here, so it's expected in the uv configuration

```
# Enable or disable automatic `sync` after `add` and `remove`.  This defaults
# to `true` when uv is enabled and `false` otherwise.
autosync = true
```

https://rye-up.com/guide/config/#config-file

**Comment by bluss** [comment]
I guess you can retitle and rewrite the issue. Best to make it clear and focused. Save the lockfile and make a reproducing example

**Comment by Zaloog** [comment]
@bluss rewrote and updated the description + screenshot

heres the content of the `requirements.lock`

```ini
# generated by rye
# use `rye lock` or `rye sync` to update this lockfile
#
# last locked with the following flags:
#   pre: false
#   features: []
#   all-features: false
#   with-sources: false

-e file:.
pandas==2.2.1
    # via pandas-test
python-dateutil==2.9.0.post0
    # via pandas
pytz==2024.1
    # via pandas
six==1.16.0
    # via python-dateutil
tzdata==2024.1
    # via pandas
```

**Comment by charliermarsh** [comment]
Unless I'm misreading it, I don't think Pandas actually declares a NumPy requirement on Python 3.11.8?

```
Requires-Dist: numpy<2,>=1.22.4; python_version < "3.11"
Requires-Dist: numpy<2,>=1.23.2; python_version == "3.11"
Requires-Dist: numpy<2,>=1.26.0; python_version >= "3.12"
```

**Comment by charliermarsh** [comment]
Oh sorry, I guess `python_version` is only major/minor.

**Comment by bluss** [comment]
It reproduces here and it doesn't install numpy


```python
>>> import pandas as pd
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/user/ryes/project/.venv/lib/python3.11/site-packages/pandas/__init__.py", line 19, in <module>
    raise ImportError(
ImportError: Unable to import required dependencies:
numpy: No module named 'numpy'
```

```
rye 0.28.0
commit: 0.28.0 (c44dc7ec8 2024-03-07)
platform: linux (x86_64)
self-python: cpython@3.12.2
symlink support: true
uv enabled: true
```

uv is 0.1.15:

```
ls ~/.rye/uv/
0.1.15/
```

**Comment by charliermarsh** [comment]
I can't reproduce with uv itself:

```
❯ echo "pandas" | uv pip compile -
Resolved 6 packages in 121ms
# This file was autogenerated by uv via the following command:
#    uv pip compile -
numpy==1.26.4
    # via pandas
pandas==2.2.1
python-dateutil==2.9.0.post0
    # via pandas
pytz==2024.1
    # via pandas
six==1.16.0
    # via python-dateutil
tzdata==2024.1
    # via pandas
```

**Comment by bluss** [comment]
rye 0.29 "ships with uv 0.1.17" supposedly but I think it doesn't update it in the self updater maybe

**Comment by bluss** [comment]
@charliermarsh  in uv itself

```
echo -e 'pandas>=2.2.1' | ~/.rye/uv/0.1.17/uv pip compile -p 3.11.8 -
Resolved 5 packages in 14ms
# This file was autogenerated by uv via the following command:
#    uv pip compile -p 3.11.8 -
pandas==2.2.1
python-dateutil==2.9.0.post0
    # via pandas
pytz==2024.1
    # via pandas
six==1.16.0
    # via python-dateutil
tzdata==2024.1
    # via pandas
```

**Comment by charliermarsh** [comment]
https://github.com/astral-sh/uv/issues/2392

**Comment by charliermarsh** [comment]
Should be fixed in uv v0.1.18 (which just went out).

**Comment by Zaloog** [comment]
@charliermarsh 
Awesome, thanks for your work and the quick fix :) 

Have a nice day
